bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjectorImpl.java

```diff
@@ -0,0 +1,123 @@
+import org.apache.felix.scr.annotations.Component;
+import org.apache.felix.scr.annotations.Property;
+import org.apache.felix.scr.annotations.Reference;
+import org.apache.felix.scr.annotations.Service;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ValueMap;
+import org.apache.sling.api.wrappers.ValueMapDecorator;
+import org.apache.sling.models.spi.DisposalCallbackRegistry;
+import org.apache.sling.models.spi.Injector;
+import org.osgi.framework.Constants;
+
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+@Component
+@Service
+@Property(name = Constants.SERVICE_RANKING, intValue = 4500)
+public class SharedValueMapValueInjectorImpl implements Injector {
+    @Reference
+    private PageRootProvider pageRootProvider;
+
+    @Override
+    public String getName() {
+        return "shared-component-properties-valuemap";
+    }
+
+    @Override
+    public Object getValue(Object adaptable, String name, Type declaredType, AnnotatedElement element, DisposalCallbackRegistry callbackRegistry) {
+        // sanity check
+        if (element.getAnnotation(SharedValueMapValue.class) == null) {
+            return null;
+        }
+
+        // sanity check
+        if (!(adaptable instanceof Resource || adaptable instanceof SlingHttpServletRequest)) {
+            return null;
+        }
+
+        Resource resource = null;
+        if (adaptable instanceof SlingHttpServletRequest) {
+            resource = ((SlingHttpServletRequest) adaptable).getResource();
+        } else if (adaptable instanceof Resource) {
+            resource = (Resource) adaptable;
+        }
+
+        if (resource != null) {
+            Page pageRoot = pageRootProvider.getRootPage(resource);
+            if (pageRoot != null) {
+                switch (element.getAnnotation(SharedValueMapValue.class).type()) {
+                    case MERGED:
+                        return getMergedProperties(pageRoot, resource).get(name);
+                    case SHARED:
+                        return getSharedProperties(pageRoot, resource).get(name);
+                    case GLOBAL:
+                        return getGlobalProperties(pageRoot, resource).get(name);
+                }
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Get global properties resource for the current resource.
+     */
+    protected ValueMap getGlobalProperties(Page pageRoot, Resource resource) {
+        String globalPropsPath = pageRoot.getPath() + "/" + JcrConstants.JCR_CONTENT + "/" + SharedComponentProperties.NN_GLOBAL_COMPONENT_PROPERTIES;
+        Resource globalPropsResource = resource.getResourceResolver().getResource(globalPropsPath);
+        return globalPropsResource != null ? globalPropsResource.getValueMap() : new ValueMapDecorator(Collections.emptyMap());
```

Comment:
You might want to change `new ValueMapDecorator(Collections.empty())` to `ValueMapDecorator.EMPTY`. Slightly cheaper.

Reply:
For sure - didnt realize that exists :)  Will do.